[Quant] Implemented 4 bit embedding op support; added corresponding test case

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69846
* #69806
* #69769
* __->__ #69768

Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

Differential Revision: [D33152673](https://our.internmc.facebook.com/intern/diff/D33152673)

dzdang committed:
[Quant] Implemented 4 bit embedding op support; added corresponding test case

Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Implemented 4 bit embedding op support; added corresponding test case"


Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3691,6 +3686,38 @@ def test_embedding_byte(self, num_embeddings, embedding_dim):
         torch.testing.assert_close(ref, qresult, atol=0.005, rtol=1e-3)
 
+    """ Tests the correctness of the quantized 4 bit embedding lookup operator """
+    @given(num_embeddings=st.integers(10, 100),
```
is it possible to merge this test with test_embedding_byte?
also we don't want to use hypothesis for testing, might be good to remove hypothesis here as well and change it to simple for loops. (remove @given, example: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_op.py#L1430), can be done in a separate PR.

dzdang replied:
Do you mean removing hypothesis for both test_embedding_byte and test_embedding_4bit or just the latter? It was already there in test_embedding_byte, so I thought I was supposed to add it. 

And yeah, I can easily combine test_embedding_byte and test_embedding_4bit. Is it acceptable to rename the function to something like test_embedding?

jerryzh168 replied:
> Do you mean removing hypothesis for both test_embedding_byte and test_embedding_4bit or just the latter? It was already there in test_embedding_byte, so I thought I was supposed to add it.

we can combine the tests and remove hypothesis, it can be in a separate PR

dzdang replied:
Sounds good. I'll put that in PR5

jerryzh168 reviewed (approved):
looks good, had a comment for test

dzdang committed:
Update on "[Quant] Implemented 4 bit embedding op support; added corresponding test case"


Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Implemented 4 bit embedding op support; added corresponding test case"


Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

jerryzh168 commented:
please import the PR with ghimport

dzdang committed:
Update on "[Quant] Implemented 4 bit embedding op support; added corresponding test case"


Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

dzdang committed:
Update on "[Quant] Implemented 4 bit embedding op support; added corresponding test case"


Summary: Support for the 4 embedding operator has been added. The support is analogous to the preexisting support for byte/8bit embedding. A corresponding test case was added to test_quantized_embedding_op.py

Reviewers: jerryzh168

Subscribers: jerryzh168, supriyar

Test plan: In pytorch main dir, execute
```
python test/test_quantization.py TestStaticQuantizedModule.test_embedding_api
```
to run the series of tests, including the newly added test_embedding_4bit
function

Tasks: T106931792

Tags

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33152673).
